### ISSUE TITLE
Broken possibility to use a proxy for SSL/WSS connections

### ISSUE THREAD

**Issue by a-zink** [eclipse-paho/paho.mqtt.java#573]
- [X] Bug exists Release Version 1.2.0 ( Master Branch)
- [X] Bug exists in MQTTv3 Client on Snapshot Version 1.2.1-SNAPSHOT (Develop Branch)
- [X] Bug exists in MQTTv5 Client on Snapshot Version 1.2.1-SNAPSHOT (Develop Branch)

In the past it was possible to set a custom SocketFactory in order to create a proxy-aware socket. However, this is broken for SSL/WSS since 1.2.0 due to SNI support introduced with https://github.com/eclipse/paho.mqtt.java/commit/bbd6dc730cc88b09226086998735cfc5867582fc

In case of a SSLSocketFactory, the TCPNetworkModule does not create the Socket using the SocketFactory anymore, and directly constructs a Socket by calling it's empty constructor ([TCPNetworkModule.java#L79](https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/TCPNetworkModule.java#L79)). 
On the next line, the Socket's connect method is called. In a network behind a proxy, this then throws an exception of course. Hence the possibility to use a proxy for SSL/WSS connections is broken.

**I'll provide a pull request.**

Please note that I have seen https://github.com/eclipse/paho.mqtt.java/pull/456, but I don't consider it as proper fix. First of all I don't like letting the code run into the exception first. This is bad style. There is also no reason to create the Socket over the plain constructor. Additionally I'm not sure if the TCPNetworkModule should check for a SSLSocketFactory. This should be handled in SSLNetworkModule or WebSocketSecureNetworkModule.

**Comment by a-zink** [comment]
This is an example for a custom, http-proxy-aware socket factory which works together with the fix. The default SSLSocketFactory (javax.net.ssl.SSLSocketFactory.getDefault()) could be used as delegate.

```
public class ProxyAwareSSLSocketFactory extends SSLSocketFactory {

	private final SSLSocketFactory delegateFactory;
	private final String proxyHost;
	private final int proxyPort;

	public ProxyAwareSSLSocketFactory(SSLSocketFactory delegateFactory, String proxyHost, int proxyPort) {
		this.delegateFactory = delegateFactory;
		this.proxyHost = proxyHost;
		this.proxyPort = proxyPort;
	}

	@Override
	public Socket createSocket() throws IOException {
		Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort));
		return new Socket(proxy);
	}

	@Override
	public Socket createSocket(Socket s, String host, int port, boolean autoClose) throws IOException {
		return delegateFactory.createSocket(s, host, port, autoClose);
	}
```

**Comment by icraggs** [comment]
I've added a fix to the develop branch, which basically reverts the previous SNI changes and replaces them with another implementation.  I haven't yet tested that the proxy support works, I hope to add a CI test which does verify that.

**Comment by a-zink** [comment]
I can get things running with your commit. However, the proxy-aware `SocketFactory` I have to provide is quite complicated compared to the simpler one I provided above for my pull request.

I would still propose to let the `SSLNetworkModule` call
`socket = ((SSLSocketFactory)factory).createSocket(socket, host, port, true);`
as I did it in my pull-request.

In summary the logic would be
- `TCPNetworkModule` calls `factory.createSocket()` and in the custom proxy-aware `SocketFactory` the `new Socket(proxy)` constructor is used.
- `TCPNetworkModule` calls `socket.connect(sockaddr)` to establish the connection over the proxy
- `SSLNetworkModule` calls `socket = ((SSLSocketFactory)factory).createSocket(socket, host, port, true)` to wrap the connected socket with a `SSLSocket`

This allows to use a simple proxy-aware `SocketFactory` like I provided above.

**Comment by icraggs** [comment]
I'll have to think about that because it's not going to work (as my testing showed) if the factory used to create the socket in TCPNetworkModule is already an SSLSocketFactory.

One way for this to work is for the SSLNetworkModule start() method not to call super.start(), and initialize its own TCP socket first in the way you suggest, but I'm not sure that'll work for your proxy approach - I would have to try it.

**Comment by a-zink** [comment]
Hm, I tested my proposal without proxy and you are right, this then throws an excpetion

**Comment by icraggs** [comment]
Released in 1.2.1

**Comment by sameerhr84** [comment]
Hi,

I have tried to update paho to 1.2.1 and tried the proxy configuration as explained here in ProxyAwareSSLSocketFactory . However I get belo exception. Any pointers?

Its working fine when I do not use proxy.

java.lang.ClassCastException: java.net.Socket cannot be cast to javax.net.ssl.SSLSocket
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(SSLNetworkModule.java:141)
	at org.eclipse.paho.client.mqttv3.internal.websocket.WebSocketSecureNetworkModule.start(WebSocketSecureNetworkModule.java:63)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:722)

**Comment by a-zink** [comment]
@sameerhr84 the provided ProxyAwareSSLSocketFactory only works with the fix I provided, but it did not make it into the release.
The following custom SocketFactory worked for me with version 1.2.1. However, as already stated above, I don't feel like this is a good solution. I would still propose a better native library support for proxies #575

```
public class ProxySocketFactory extends SSLSocketFactory {
	private final SSLSocketFactory delegate;
	private final Proxy proxy;

	public ProxySocketFactory() {
		this((SSLSocketFactory) SSLSocketFactory.getDefault(), resolveProxyHost(), resolveProxyPort());
	}

	private static Integer resolveProxyPort() {
		return Integer.valueOf(System.getProperty("http.proxyPort", "3128"));
	}

	private static String resolveProxyHost() {
		return System.getProperty("http.proxyHost", "localhost");
	}

	public ProxySocketFactory(SSLSocketFactory delegate, String proxyHost, int proxyPort) {
		this.delegate = delegate;
		this.proxy = buildProxy(proxyHost, proxyPort);
	}

	private Proxy buildProxy(String proxyHost, int proxyPort) {
		return new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort));
	}

	@Override
	public Socket createSocket() throws IOException {
		return new ProxiedSSLSocket(delegate, new Socket(proxy));
	}

	@Override
	public String[] getDefaultCipherSuites() {
		return delegate.getDefaultCipherSuites();
	}

	@Override
	public String[] getSupportedCipherSuites() {
		return delegate.getSupportedCipherSuites();
	}

	@Override
	public Socket createSocket(Socket s, String host, int port, boolean autoClose) throws IOException {
		return delegate.createSocket(s, host, port, autoClose);
	}

	@Override
	public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
		return delegate.createSocket(host, port);
	}

	@Override
	public Socket createSocket(String host, int port, InetAddress localHost, int localPort)
			throws IOException, UnknownHostException {
		return delegate.createSocket(host, port, localHost, localPort);
	}

	@Override
	public Socket createSocket(InetAddress host, int port) throws IOException {
		return delegate.createSocket(host, port);
	}

	@Override
	public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort)
			throws IOException {
		return delegate.createSocket(address, port, localAddress, localPort);
	}

}
```

```
public class ProxiedSSLSocket extends SSLSocket {
	private final SSLSocketFactory socketFactory;
	private final Socket proxySocket;
	private SSLSocket socket;

	public ProxiedSSLSocket(SSLSocketFactory socketFactory, Socket proxySocket) {
		this.socketFactory = socketFactory;
		this.proxySocket = proxySocket;
	}

	@Override
	public void connect(SocketAddress socketAddress) throws IOException {
		connect(socketAddress, 0);
	}

	@Override
	public void connect(SocketAddress socketAddress, int timeout) throws IOException {
		proxySocket.connect(socketAddress, timeout);
		this.socket = (SSLSocket) socketFactory.createSocket(proxySocket,
				((InetSocketAddress) socketAddress).getHostName(), ((InetSocketAddress) socketAddress).getPort(), true);
	}

	@Override
	public String[] getSupportedCipherSuites() {
		return this.socket.getSupportedCipherSuites();
	}

	@Override
	public String[] getEnabledCipherSuites() {
		return this.socket.getEnabledCipherSuites();
	}

	@Override
	public void setEnabledCipherSuites(String[] strings) {
		this.socket.setEnabledCipherSuites(strings);
	}

	@Override
	public String[] getSupportedProtocols() {
		return this.socket.getSupportedProtocols();
	}

	@Override
	public String[] getEnabledProtocols() {
		return this.socket.getEnabledProtocols();
	}

	@Override
	public void setEnabledProtocols(String[] strings) {
		this.socket.setEnabledProtocols(strings);
	}

	@Override
	public SSLSession getSession() {
		return this.socket.getSession();
	}

	@Override
	public SSLSession getHandshakeSession() {
		return this.socket.getHandshakeSession();
	}

	@Override
	public void addHandshakeCompletedListener(HandshakeCompletedListener handshakeCompletedListener) {
		this.socket.addHandshakeCompletedListener(handshakeCompletedListener);
	}

	@Override
	public void removeHandshakeCompletedListener(HandshakeCompletedListener handshakeCompletedListener) {
		this.socket.removeHandshakeCompletedListener(handshakeCompletedListener);
	}

	@Override
	public void startHandshake() throws IOException {
		this.socket.startHandshake();
	}

	@Override
	public void setUseClientMode(boolean b) {
		this.socket.setUseClientMode(b);
	}

	@Override
	public boolean getUseClientMode() {
		return this.socket.getUseClientMode();
	}

	@Override
	public void setNeedClientAuth(boolean b) {
		this.socket.setNeedClientAuth(b);
	}

	@Override
	public boolean getNeedClientAuth() {
		return this.socket.getNeedClientAuth();
	}

	@Override
	public void setWantClientAuth(boolean b) {
		this.socket.setWantClientAuth(b);
	}

	@Override
	public boolean getWantClientAuth() {
		return this.socket.getWantClientAuth();
	}

	@Override
	public void setEnableSessionCreation(boolean b) {
		this.socket.setEnableSessionCreation(b);
	}

	@Override
	public boolean getEnableSessionCreation() {
		return this.socket.getEnableSessionCreation();
	}

	@Override
	public SSLParameters getSSLParameters() {
		return this.socket.getSSLParameters();
	}

	@Override
	public void setSSLParameters(SSLParameters sslParameters) {
		this.socket.setSSLParameters(sslParameters);
	}

	@Override
	public void bind(SocketAddress socketAddress) throws IOException {
		throw new UnsupportedOperationException();
	}

	@Override
	public InetAddress getInetAddress() {
		return this.socket.getInetAddress();
	}

	@Override
	public InetAddress getLocalAddress() {
		return this.socket.getLocalAddress();
	}

	@Override
	public int getPort() {
		return this.socket.getPort();
	}

	@Override
	public int getLocalPort() {
		return this.socket.getLocalPort();
	}

	@Override
	public SocketAddress getRemoteSocketAddress() {
		return this.socket.getRemoteSocketAddress();
	}

	@Override
	public SocketAddress getLocalSocketAddress() {
		return this.socket.getLocalSocketAddress();
	}

	@Override
	public SocketChannel getChannel() {
		return this.socket.getChannel();
	}

	@Override
	public InputStream getInputStream() throws IOException {
		return this.socket.getInputStream();
	}

	@Override
	public OutputStream getOutputStream() throws IOException {
		return this.socket.getOutputStream();
	}

	@Override
	public void setTcpNoDelay(boolean b) throws SocketException {
		this.socket.setTcpNoDelay(b);
	}

	@Override
	public boolean getTcpNoDelay() throws SocketException {
		return this.socket.getTcpNoDelay();
	}

	@Override
	public void setSoLinger(boolean b, int i) throws SocketException {
		this.socket.setSoLinger(b, i);
	}

	@Override
	public int getSoLinger() throws SocketException {
		return this.socket.getSoLinger();
	}

	@Override
	public void sendUrgentData(int i) throws IOException {
		this.socket.sendUrgentData(i);
	}

	@Override
	public void setOOBInline(boolean b) throws SocketException {
		this.socket.setOOBInline(b);
	}

	@Override
	public boolean getOOBInline() throws SocketException {
		return this.socket.getOOBInline();
	}

	@Override
	public void setSoTimeout(int i) throws SocketException {
//		this.socket.setSoTimeout(i);
	}

	@Override
	public int getSoTimeout() throws SocketException {
		return this.socket.getSoTimeout();
	}

	@Override
	public void setSendBufferSize(int i) throws SocketException {
		this.socket.setSendBufferSize(i);
	}

	@Override
	public int getSendBufferSize() throws SocketException {
		return this.socket.getSendBufferSize();
	}

	@Override
	public void setReceiveBufferSize(int i) throws SocketException {
		this.socket.setReceiveBufferSize(i);
	}

	@Override
	public int getReceiveBufferSize() throws SocketException {
		return this.socket.getReceiveBufferSize();
	}

	@Override
	public void setKeepAlive(boolean b) throws SocketException {
		this.socket.setKeepAlive(b);
	}

	@Override
	public boolean getKeepAlive() throws SocketException {
		return this.socket.getKeepAlive();
	}

	@Override
	public void setTrafficClass(int i) throws SocketException {
		this.socket.setTrafficClass(i);
	}

	@Override
	public int getTrafficClass() throws SocketException {
		return this.socket.getTrafficClass();
	}

	@Override
	public void setReuseAddress(boolean b) throws SocketException {
		this.socket.setReuseAddress(b);
	}

	@Override
	public boolean getReuseAddress() throws SocketException {
		return this.socket.getReuseAddress();
	}

	@Override
	public void close() throws IOException {
		this.socket.close();
	}

	@Override
	public void shutdownInput() throws IOException {
		this.socket.shutdownInput();
	}

	@Override
	public void shutdownOutput() throws IOException {
		this.socket.shutdownOutput();
	}

	@Override
	public String toString() {
		return this.socket.toString();
	}

	@Override
	public boolean isConnected() {
		return this.socket.isConnected();
	}

	@Override
	public boolean isBound() {
		return this.socket.isBound();
	}

	@Override
	public boolean isClosed() {
		return this.socket.isClosed();
	}

	@Override
	public boolean isInputShutdown() {
		return this.socket.isInputShutdown();
	}

	@Override
	public boolean isOutputShutdown() {
		return this.socket.isOutputShutdown();
	}

	@Override
	public void setPerformancePreferences(int i, int i1, int i2) {
		this.socket.setPerformancePreferences(i, i1, i2);
	}
}
```

**Comment by sameerhr84** [comment]
@a-zink  I had to make this working with a similar fix being a critical issue for the project. I also agree with you that the right fix is to fix #575. Any way to take this on priority?

**Comment by AndrewJudson** [comment]
@a-zink that code works for me for a non authenticating proxy. Do you know what we would need to do to make it work with an authenticating proxy? The Authenticator gets set in my app which makes it work with the proxy for http requests but I seem to still get a 407 authentication required error here.

**Comment by a-zink** [comment]
I have not tried this with `java.net.Authenticator`, but I could try when I find some time.

Every tool usually has it's own approach to deal with proxy auth, or in worst case it is missing at all. Additionally your credentials get spread accross your system over the tool configs. Hence I tend to solve proxy auth by using tools like [cntlm](http://cntlm.sourceforge.net/) or similar.
This is a local proxy which e.g. runs on localhost:3128 and is configured against your corporate proxy + auth credentials. You then use localhost:3128 as proxy in your tools/code and don't need to handle auth by yourself. 
This might sound cumbersome, but is actually quite handy if you have to change your password frequently, because you only need to update one single tool.

**Comment by sameerhr84** [comment]
@a-zink, I am not able to use 1.2.1 due to the issue https://github.com/eclipse/paho.mqtt.java/issues/637
Does your fix work with 1.2.0
